### PR TITLE
Add handover features

### PR DESCRIPTION
**Pull request by cbranch**
This changeset adds the capability for the old process to send runtime-generated data to the new process.

There are two forms:

1. Environment variables, set when configuring restart behaviour, so are static for the lifetime of the program; this is appropriate for passing data/resources created on startup
2. Stream data, sent after spawning the new process; this allows state to be serialised between processes

State transfer is intentionally unidirectional from old to new. The pipe is represented as a boxed AsyncRead/Write. This API is intentionally quite simple; e.g. supporting file descriptor passing is not a goal in this version, we should probably consider a better API for that than exposing e.g. a RawFd and letting users go wild.

### CODE REVIEW

**Review by nmldiegues — COMMENTED** [review]
Nit:  unit tests are not compiling in this repo, should be an easy fix

File: src/lib.rs [hunk 1]
```diff
@@ -59,23 +78,29 @@ pub type RestartResult<T> = anyhow::Result<T>;
...
-#[derive(Clone, Debug, Deserialize)]
 pub struct RestartConfig {
     /// Enables the restart coordination socket for graceful restarts as an alternative to the SIGUSR1 signal.
     pub enabled: bool,
     /// Socket path
     pub coordination_socket_path: PathBuf,
+    /// Sets environment variables on the newly-started process
+    pub environment: Vec<(OsString, OsString)>,
+    /// Receive fine-grained events on the lifecycle of the new process and support data transfer.
+    pub lifecycle_handler: Box<dyn LifecycleHandler>,
```
[the remark this answers is not in the document]
Comment (nox) [reply src/lib.rs:94]: Whether or not we implement `Default` here, as this is a public struct, adding those fields is a breaking change and that should be released as shellflip 2.

File: src/lib.rs [hunk 2]
```diff
@@ -348,17 +398,27 @@ fn clear_cloexec(fd: RawFd) -> nix::Result<()> {
 }
 
 /// Attempt to start a new instance of this proxy.
-fn spawn_child(restart_fd: Option<RawFd>) -> io::Result<process::Child> {
+async fn spawn_child(
```
Comment (nox) [inline src/lib.rs:401]: Btw this function should probably be unsafe, as callers need to make sure `restart_fd` isn't a fd that would be unsound to pass to the child.

File: src/pipes.rs [hunk 3]
```diff
@@ -71,4 +80,27 @@ impl CompletionSender {
...
+
+pub(crate) trait FdStringExt {
+    fn fd_string(&self) -> String;
+    fn from_fd_string(fd_str: &str) -> io::Result<Self>
+    where
+        Self: Sized;
+}
+
+impl<T: AsRawFd + FromRawFd> FdStringExt for T {
+    fn from_fd_string(fd_str: &str) -> io::Result<Self> {
```
Comment (nox) [inline src/pipes.rs:93]: This def should be unsafe.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cbranch) [reply src/lib.rs:94]: There's a `Default` impl below it (which is what this should have had before!)
Side note, should the major version bump happen *here* or in a followup commit?

[the remark this answers is not in the document]
Comment (nox) [reply src/lib.rs:94]: I guess in a separate commit after all the changes, as there are already multiple commits in the PR. Whichever you prefer really.

File: src/lib.rs (hunk 2 above)
Comment (cbranch) [reply src/lib.rs:405]: I see your point, but since this fd is created by us, we can wrap it appropriately to ensure safety through the type system.

**Review by nmldiegues — APPROVED** [review]

**Pull request merged**